egs/wsj/s5/steps/nnet3/libs/xconfig_layers.py

```diff
@@ -0,0 +1,679 @@
+from __future__ import print_function
+import subprocess
+import logging
+import math
+import re
+import sys
+import traceback
+import time
+import argparse
+import xconfig_utils
+
+
+# A base-class for classes representing layers of xconfig files.
+# This mainly just sets self.layer_type, self.name and self.config,
+class XconfigLayerBase(object):
+    # Constructor.
+    # first_token is the first token on the xconfig line, e.g. 'affine-layer'.f
+    # key_to_value is a dict like:
+    # { 'name':'affine1', 'input':'Append(0, 1, 2, ReplaceIndex(ivector, t, 0))', 'dim=1024' }.
+    # The only required and 'special' values that are dealt with directly at this level, are
+    # 'name' and 'input'.
+    # The rest are put in self.config and are dealt with by the child classes' init functions.
+    # all_layers is an array of objects inheriting XconfigLayerBase for all previously
+    # parsed layers.
+
+    def __init__(self, first_token, key_to_value, all_layers):
+        self.layer_type = first_token
+        if not 'name' in key_to_value:
+            raise RuntimeError("Expected 'name' to be specified.")
+        self.name = key_to_value['name']
+        if not xconfig_utils.IsValidLineName(self.name):
+            raise RuntimeError("Invalid value: name={0}".format(key_to_value['name']))
+
+        # the following, which should be overridden in the child class, sets
+        # default config parameters in self.config.
+        self.SetDefaultConfigs()
+        # The following is not to be reimplemented in child classes;
+        # it sets the config values to those specified by the user, and
+        # parses any Descriptors.
+        self.SetConfigs(key_to_value, all_layers)
+        # the following, which should be overridden in the child class, checks
+        # that the config parameters that have been set are reasonable.
+        self.CheckConfigs()
+
+
+    # We broke this code out of __init__ for clarity.
+    def SetConfigs(self, key_to_value, all_layers):
+        # the child-class constructor will deal with the configuration values
+        # in a more specific way.
+        for key,value in key_to_value.items():
+            if key != 'name':
+                if not key in self.config:
+                    raise RuntimeError("Configuration value {0}={1} was not expected in "
+                                    "layer of type {2}".format(key, value, self.layer_type))
+                self.config[key] = xconfig_utils.ConvertValueToType(key, type(self.config[key]), value)
+
+
+        self.descriptors = dict()
+        self.descriptor_dims = dict()
+        # Parse Descriptors and get their dims and their 'final' string form.
+        # Put them as 4-tuples (descriptor, string, normalized-string, final-string)
+        # in self.descriptors[key]
+        for key in self.GetDescriptorConfigs():
```

Comment:
This function just returns the names of the input descriptors used by a layer, right ? Or do you plan to use it to store other things ?

I would like to call it `GetInputDescriptorNames()` as it will be just returning the expected input names and not the actual configs.